File: packages/sdk-ts/src/circuitBreaker.ts
export interface CircuitBreakerConfig {
  maxRepeatedCalls?: number; // Max identical tool calls allowed in window
  timeWindowMs?: number;      // Time window in ms (e.g., 10000ms)
}

export class CircuitBreaker {
  private callTracker: Map<string, { paramsHash: string; timestamp: number }[]> = new Map();
  private trippedBreakers: Map<string, number> = new Map(); // sessionKey -> resetTimestamp

  check(sessionKey: string, toolName: string, params: Record<string, any>, config: CircuitBreakerConfig): { tripped: boolean; reason?: string } {
    const maxCalls = config.maxRepeatedCalls || 4;
    const windowMs = config.timeWindowMs || 10000;
    const now = Date.now();

    // Check if circuit breaker is currently open (tripped)
    const resetTime = this.trippedBreakers.get(sessionKey);
    if (resetTime && now < resetTime) {
      const remainingSec = Math.ceil((resetTime - now) / 1000);
      return {
        tripped: true,
        reason: `Circuit Breaker OPEN: Agent loop detected. Aborting tool calls for next ${remainingSec}s.`,
      };
    }

    const trackerKey = `${sessionKey}:${toolName}`;
    const paramsHash = JSON.stringify(params);

    const history = (this.callTracker.get(trackerKey) || []).filter(item => now - item.timestamp < windowMs);

    // Check for identical repeated tool calls
    const repeatedCount = history.filter(item => item.paramsHash === paramsHash).length;
    if (repeatedCount >= maxCalls) {
      // Trip the breaker for 30 seconds
      this.trippedBreakers.set(sessionKey, now + 30000);
      return {
        tripped: true,
        reason: `Circuit Breaker TRIPPED: Tool '${toolName}' called ${repeatedCount + 1} times with identical parameters within ${windowMs / 1000}s loop.`,
      };
    }

    history.push({ paramsHash, timestamp: now });
    this.callTracker.set(trackerKey, history);

    return { tripped: false };
  }

  reset(sessionKey: string): void {
    this.trippedBreakers.delete(sessionKey);
  }
}

File: packages/sdk-java/src/main/java/dev/agentshield/CircuitBreaker.java
package dev.agentshield;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.List;
import java.util.ArrayList;
import java.util.stream.Collectors;

public class CircuitBreaker {

    private static class CallRecord {
        String paramsHash;
        long timestamp;

        CallRecord(String paramsHash, long timestamp) {
            this.paramsHash = paramsHash;
            this.timestamp = timestamp;
        }
    }

    private final Map<String, List<CallRecord>> callTracker = new ConcurrentHashMap<>();
    private final Map<String, Long> trippedBreakers = new ConcurrentHashMap<>();

    public CircuitBreakerResult check(String sessionKey, String toolName, Map<String, Object> params, GuardrailPolicy.CircuitBreakerConfig config) {
        int maxCalls = config != null && config.getMaxRepeatedCalls() != null ? config.getMaxRepeatedCalls() : 4;
        int windowMs = config != null && config.getTimeWindowMs() != null ? config.getTimeWindowMs() : 10000;
        long now = System.currentTimeMillis();

        // Check if circuit breaker is currently open (tripped)
        Long resetTime = trippedBreakers.get(sessionKey);
        if (resetTime != null && now < resetTime) {
            long remainingSec = (resetTime - now + 999) / 1000;
            return new CircuitBreakerResult(true, "Circuit Breaker OPEN: Agent loop detected. Aborting tool calls for next " + remainingSec + "s.");
        }

        String trackerKey = sessionKey + ":" + toolName;
        String paramsHash = hashParams(params);

        List<CallRecord> history = callTracker.getOrDefault(trackerKey, new ArrayList<>());

        // Filter recent calls within time window
        List<CallRecord> recent = history.stream()
            .filter(record -> now - record.timestamp < windowMs)
            .collect(Collectors.toList());

        // Check for identical repeated tool calls
        long repeatedCount = recent.stream()
            .filter(record -> record.paramsHash.equals(paramsHash))
            .count();

        if (repeatedCount >= maxCalls) {
            // Trip the breaker for 30 seconds
            trippedBreakers.put(sessionKey, now + 30000);
            return new CircuitBreakerResult(true,
                "Circuit Breaker TRIPPED: Tool '" + toolName + "' called " + (repeatedCount + 1) + " times with identical parameters within " + (windowMs / 1000) + "s loop.");
        }

        recent.add(new CallRecord(paramsHash, now));
        callTracker.put(trackerKey, recent);

        return new CircuitBreakerResult(false, null);
    }

    public void reset(String sessionKey) {
        trippedBreakers.remove(sessionKey);
    }

    private String hashParams(Map<String, Object> params) {
        try {
            String json = new com.fasterxml.jackson.databind.ObjectMapper().writeValueAsString(params);
            long hash = 1469598103934665603L;
            for (byte b : json.getBytes()) {
                hash ^= b & 0xFF;
                hash *= 1099511628211L;
            }
            return Long.toHexString(hash);
        } catch (Exception e) {
            return params.toString();
        }
    }

    public static class CircuitBreakerResult {
        private boolean tripped;
        private String reason;

        public CircuitBreakerResult(boolean tripped, String reason) {
            this.tripped = tripped;
            this.reason = reason;
        }

        public boolean isTripped() { return tripped; }
        public String getReason() { return reason; }
    }
}